packages/vike-vue/src/integration/onRenderHtml.ts

```diff
@@ -143,17 +145,17 @@ async function getHeadHtml(pageContext: PageContextServer & PageContextInternal)
   return headHtml
 }
 
-async function getBodyHtmlBoundary(pageContext: PageContextServer) {
-  const bodyHtmlBegin = dangerouslySkipEscape(
-    (await callCumulativeHooks(pageContext.config.bodyHtmlBegin, pageContext)).join(''),
-  )
-
+async function getHtmlInjections(pageContext: PageContextServer) {
+  const { config } = pageContext
   // we define this hook here so that it doesn't need to be exported by vike-vue
   const defaultTeleport = `<div id="teleported">${pageContext.ssrContext!.teleports?.['#teleported'] ?? ''}</div>`
-  const bodyHtmlEndHooks = [defaultTeleport, ...(pageContext.config.bodyHtmlEnd ?? [])]
-  const bodyHtmlEnd = dangerouslySkipEscape((await callCumulativeHooks(bodyHtmlEndHooks, pageContext)).join(''))
-
-  return { bodyHtmlBegin, bodyHtmlEnd }
+  const [headHtmlBegin, headHtmlEnd, bodyHtmlBegin, bodyHtmlEnd] = await Promise.all([
+    dangerouslySkipEscape((await callCumulativeHooks(config.headHtmlBegin, pageContext)).join('')),
+    dangerouslySkipEscape([defaultTeleport, ...(await callCumulativeHooks(config.headHtmlEnd, pageContext))].join('')),
+    dangerouslySkipEscape((await callCumulativeHooks(config.bodyHtmlBegin, pageContext)).join('')),
+    dangerouslySkipEscape((await callCumulativeHooks(config.bodyHtmlEnd, pageContext)).join('')),
+  ])
```

Comment:
Slight code improvements, I found it pretty hard to read at first.
Also, the call to `Promise.all` isn't quite proper here, since we're still calling await at the top level, so we're not actually creating any promises. But the idea is correct I think, we can call all the hooks in parallel here by modifying this code a little.

```suggestion
  const renderHooks = async (hooks: unknown[] | undefined) => {
    const values = await callCumulativeHooks(hooks, pageContext)
    return dangerouslySkipEscape(values.join(''))
  }

  const [headHtmlBegin, headHtmlEnd, bodyHtmlBegin, bodyHtmlEnd] = await Promise.all([
    renderHooks(config.headHtmlBegin),
    renderHooks(config.headHtmlEnd),
    renderHooks(config.bodyHtmlBegin),
    renderHooks([defaultTeleport,  ...(config.bodyHtmlEnd ?? [])]),
  ])
```